sw/device/tests/rv_timer_systick_test.c

```diff
@@ -0,0 +1,93 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_rv_timer.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/irq.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+static dif_rv_timer_t timer;
+
+enum {
+  kHart = kTopEarlgreyPlicTargetIbex0,
+  kComparator = 0,
+  kReferenceTimeMicros = 1000 * 1000,  // 1 second
+};
+
+OTTF_DEFINE_TEST_CONFIG();
+
+status_t set_tick(uint32_t tick_hz) {
+  dif_rv_timer_tick_params_t tick_params;
+  TRY(dif_rv_timer_approximate_tick_params(kClockFreqPeripheralHz, tick_hz,
+                                           &tick_params));
+  TRY(dif_rv_timer_set_tick_params(&timer, kHart, tick_params));
+  return OK_STATUS();
+}
+
+status_t test_tick(uint32_t tick_hz) {
+  LOG_INFO("%s: tick_hz = %u", __func__, tick_hz);
+  uint64_t counter = 0;
+
+  TRY(set_tick(tick_hz));
+
+  TRY(dif_rv_timer_counter_write(&timer, kHart, 0));
+  TRY(dif_rv_timer_counter_read(&timer, kHart, &counter));
+  TRY_CHECK(counter == 0, "Failed to write the counter");
+
+  TRY(dif_rv_timer_counter_set_enabled(&timer, kHart, kDifToggleEnabled));
+  busy_spin_micros(kReferenceTimeMicros);
+  TRY(dif_rv_timer_counter_set_enabled(&timer, kHart, kDifToggleDisabled));
+
+  TRY(dif_rv_timer_counter_read(&timer, kHart, &counter));
+  const uint64_t elapsed_micros = udiv64_slow(counter, tick_hz, NULL) * 1000000;
+
+  // Verify that `n * T ~= 1 second` within 1% of tolerance.
```

Comment:
The clocks in the AST are less accurate: they are spec'ed at +-3%. The fpga's clocks are better. You can expect 1% for kDeviceType == kDeviceFpgaCw310 (from sw/device/lib/arch/device.h). We will need to create a silicon device type, but its tolerance is expected to be similar to kDeviceSimDV.